File: sundarr/app/schemas/storage.py
from typing import Literal

from pydantic import BaseModel, Field


class StorageConfigRequest(BaseModel):
    type: Literal["smb"] = "smb"
    host: str = Field(min_length=1)
    port: int = Field(default=445, ge=1, le=65535)
    share: str = Field(min_length=1)
    username: str = Field(min_length=1)
    password: str | None = None
    domain: str = ""
    base_path: str = "/"
    libraries: dict[str, str] = Field(default_factory=dict)


class StorageConfigResponse(BaseModel):
    type: Literal["smb"] = "smb"
    host: str = ""
    port: int = 445
    share: str = ""
    username: str = ""
    password_set: bool = False
    domain: str = ""
    base_path: str = "/"
    libraries: dict[str, str] = Field(default_factory=dict)


class StorageConfigTestResponse(BaseModel):
    ok: bool
    error_code: str | None = None
    error_message: str | None = None


class StorageBrowseEntry(BaseModel):
    name: str
    path: str
    is_dir: bool
    size: int | None = None


class StorageBrowseResponse(BaseModel):
    path: str
    entries: list[StorageBrowseEntry] = Field(default_factory=list)

File: web/src/types/storage.ts
export type StorageConfigResponse = {
  type: 'smb'
  host: string
  port: number
  share: string
  username: string
  password_set: boolean
  domain: string
  base_path: string
  libraries: Record<string, string>
}

export type SmbConnectionResponse = {
  id: string
  name: string
  enabled: boolean
  host: string
  port: number
  share: string
  username: string
  password_set: boolean
  domain: string
  base_path: string
  bound_local_libraries: string[]
  bound_remote_libraries: string[]
  last_test_ok: boolean | null
  last_test_error_code: string | null
  last_test_error_message: string | null
  created_at: string | null
  updated_at: string | null
}

export type SmbConnectionListResponse = {
  count: number
  page: number
  page_size: number
  results: SmbConnectionResponse[]
}

export type StorageConfigRequest = Omit<StorageConfigResponse, 'password_set'> & {
  password: string | null
}

export type StorageConfigTestResponse = {
  ok: boolean
  error_code: string | null
  error_message: string | null
}

export type StorageBrowseResponse = {
  path: string
  entries: StorageBrowseEntry[]
}

export type StorageBrowseEntry = {
  name: string
  path: string
  is_dir: boolean
  size: number | null
  modified_at: string | null
}

export type StorageFormState = {
  host: string
  port: string
  share: string
  username: string
  password: string
  domain: string
  base_path: string
  library_movies: string
  library_tv: string
  library_anime: string
}
